lib/ui/painting/immutable_buffer.cc

```diff
@@ -49,6 +53,50 @@ void ImmutableBuffer::init(Dart_NativeArguments args) {
   tonic::DartInvoke(callback_handle, {Dart_TypeVoid()});
 }
 
+void ImmutableBuffer::initFromAsset(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  Dart_Handle callback_handle = Dart_GetNativeArgument(args, 2);
+  if (!Dart_IsClosure(callback_handle)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Callback must be a function"));
+    return;
+  }
+  Dart_Handle asset_name_handle = Dart_GetNativeArgument(args, 1);
+  uint8_t* chars = nullptr;
+  intptr_t asset_length = 0;
+  Dart_Handle result =
+      Dart_StringToUTF8(asset_name_handle, &chars, &asset_length);
+  if (Dart_IsError(result)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Asset must be valid UTF8"));
+    return;
+  }
+  Dart_Handle immutable_buffer = Dart_GetNativeArgument(args, 0);
+
+  std::string asset_name = std::string{reinterpret_cast<const char*>(chars),
+                                       static_cast<size_t>(asset_length)};
+
+  std::shared_ptr<AssetManager> asset_manager = UIDartState::Current()
+                                                    ->platform_configuration()
+                                                    ->client()
+                                                    ->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data = asset_manager->GetAsMapping(asset_name);
```

Comment:
Are these mappings thread safe? These usually come from `flutter::AssetResolver`s and looking at the inheritance diagram from these, `flutter::APKAssetProvider` seems to be a subclass. Its assets are backed by AAsset handles whose thread safety notes say "AAsset objects are NOT thread-safe, and should not be shared across threads.". Since there used to be a copy earlier, it sort of didn't matter. But now that these are referenced across thread (for texture decompression), there may be threading violations that we have perhaps not contended with.

Reply:
Hm. Sounds like we can't avoid the copy in that case.

If we didn't expose the lenght, we could get away with this being something that's done entirely on the decoder thread though right? We could just give that a way to acquire the data and then let it do the whole thing there.

Reply:
We don't need the length in the framework for this case. We could always say "length might be -1 for certain instances not created from a uint8list"

Reply:
we could also copy once into an skdata - that would still reduce the number of copies by 1. Though having a single createCodecFromAsset method would mean we could do the entire thing on the right thread, i'm not sure how well that works with the framework wanting to abstract byte acquisition from image decoding.

I don't think adding a special case for asset images would be a bad thing fwiw, but it would be a special case.